### ISSUE TITLE
Support docker-java 3.0.0

### ISSUE THREAD

**Issue by llamahunter** [bmuschko/gradle-docker-plugin#180]
It appears there are several API changes in 3.0.0. Trying to run gradle-docker-plugin using the 3.0.0rc3 build of docker-java fails at runtime due to signature mismatches.  Due to the way gradle-docker-plugin dynamically binds to docker-java, there are unfortunately no compile time errors.

In particular, the DockerClientConfig API has changed dramatically so as to better match the CLI configuration.

**Comment by cdancy** [comment]
@llamahunter we are indeed aware of this. There are no immediate plans to move to 3.0.0 but it will have to happen at some point in the not too distant future.

**Comment by llamahunter** [comment]
Ok... I first search for an existing issue on this, but didn't see one.  Thought people should be aware that it's not trivial to move to 3.0.0.

**Comment by cdancy** [comment]
@llamahunter thanks for the heads up anyway. Yeah I gave things a go with their 3.0.0 release not too long ago as that branch has some features we're interested in. Failures more/less everywhere. Package layout is different as you hit with the binds issue. It will take a bit of effort though the guys at docker-java assure me the changes should not be that drastic and should be mostly name related issues.

**Comment by llamahunter** [comment]
Is there a good reason why gradle-docker-plugin runtime binds to docker-java rather than treating it as a compile time library?  It seems that lots of things in the plugin would be easier to understand without all the reflection.

**Comment by cdancy** [comment]
Anytime you apply a plugin you do so by putting its dependencies, and transitive dependencies, on the classpath. If some other plugin does the same, but say with a higher version of some dependency you need, then gradle will choose the higher version which may break your plugin. Having a custom classloader allows us to get around this issue and ensures we don't have any potential clobbering from other dependencies.

**Comment by llamahunter** [comment]
Any movement here?  Docker released remote api version v1.23 (docker-machine v1.11), but gradle-docker-plugin is still unable to work with api v1.22 (docker-machine v1.10), as such support was added after docker-java 2.2.3 was released.

**Comment by cdancy** [comment]
Not until they release v3.0.0

**Comment by orzeh** [comment]
Taking into account @KostyaSha [comment](https://github.com/docker-java/docker-java/issues/552#issuecomment-211528597) maybe we should create a new branch which will use docker-java v3.0.0 RC and start work to align gradle-docker-plugin to new APIs?

**Comment by KostyaSha** [comment]
> Not until they release v3.0.0

We are very slow because of lack of free time. If gradle-plugin wouldn't have issues with the scope of APIs that it uses, then you should be fine with RC. We do RC because we don't want to do massive changes after and with docker APIs it not really possible.

**Comment by cdancy** [comment]
@orzeh, or anyone in this thread, I've [created a branch here](https://github.com/cdancy/gradle-docker-plugin/tree/docker-java-3) if you'd like to help out and start sending in PR's. Any help would be greatly appreciated.

**Comment by orzeh** [comment]
ok, I`ll start work on it.

Do you plan to publish plugin snapshots?

**Comment by cdancy** [comment]
@orzeh thanks! I wasn't planning on it. However, the migration _shouldn't_ be that bad and if we can knock things out relatively quick I can do another release.

**Comment by orzeh** [comment]
I've made initial PR https://github.com/cdancy/gradle-docker-plugin/pull/1

**Comment by StFS** [comment]
Just FYI, I'm getting a stack trace when pushing my docker image into an Artifactory docker repository. The task doesn't fail but the stack trace is logged as an error and it looks ugly.

I've tracked the "error" to docker-java and from what I understand there is a fix already in place in v3.0:
https://github.com/docker-java/docker-java/commit/f2954e037ef6be098b1ef838480d61421ed34ceb

Here's the error message that I'm getting when pushing my image:

```
Pushing image with name 'docker.repo.nextcode.com/gor-scripts:5.9-SNAPSHOT'.
[pool-8-thread-1] ERROR com.github.dockerjava.core.async.ResultCallbackTemplate - Error during callback
com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "aux" (class com.github.dockerjava.api.model.PushResponseItem), not marked as ignorable (9 known properties: , "stream", "error", "from", "progress", "time", "status", "errorDetail", "id", "progressDetail"])
 at [Source: N/A; line: -1, column: -1] (through reference chain: com.github.dockerjava.api.model.PushResponseItem["aux"])
	at com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException.from(UnrecognizedPropertyException.java:79)
	at com.fasterxml.jackson.databind.DeserializationContext.reportUnknownProperty(DeserializationContext.java:579)
	at com.fasterxml.jackson.databind.deser.std.StdDeserializer.handleUnknownProperty(StdDeserializer.java:672)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownProperty(BeanDeserializerBase.java:906)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:328)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:121)
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:2769)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:1478)
	at com.fasterxml.jackson.databind.ObjectMapper.treeToValue(ObjectMapper.java:1811)
	at com.github.dockerjava.core.async.JsonStreamProcessor.processResponseStream(JsonStreamProcessor.java:48)
	at com.github.dockerjava.jaxrs.async.AbstractCallbackNotifier.call(AbstractCallbackNotifier.java:62)
	at com.github.dockerjava.jaxrs.async.AbstractCallbackNotifier.call(AbstractCallbackNotifier.java:22)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

BUILD SUCCESSFUL

Total time: 1 mins 9.951 secs
```

**Comment by cdancy** [comment]
@StFS you're correct. Once we move to docker-java 3.0.0 this issue will go away. @orzeh has already put together a PR for this @ https://github.com/cdancy/gradle-docker-plugin/pull/1 which you can build/host locally and give things a go.

FYI ... we won't rebase the plugin off of docker-java 3.0.0 until they release. They are still releasing RC's, which is fine, but for sanities sake we can't move forward until they do.

**Comment by StFS** [comment]
Understood. Completely agree.

**Comment by cdancy** [comment]
FYI: docker-java-3.0.0 branch has been merged with PR #217. Assuming my testing here goes well I'll release probably Sunday night sometime. Any extra real-world testing would be appreciated.

**Comment by cdancy** [comment]
gradle-docker-plugin 3.0.0 has been released and thanks to the hardwork put in by @orzeh we've rebased off of docker-java 3.0.0.
